Mathlib/Combinatorics/Flow/Basic.lean

```diff
@@ -0,0 +1,167 @@
+def flowOut [AddCommMonoid R] (f : V → V → R) (v : V) := ∑ w, f v w
+/-- The excess flow of a vertex (the incoming flow minus the outgoing flow). --/
+def excess [SubtractionCommMonoid R] (f : V → V → R) (v : V) := flowIn f v - flowOut f v
+
+variable [OrderedAddCommMonoid R] {N : Network V R}
+
+/-- A valid assignment of flow units to all edges for a given `Network.Problem`.
+
+    Note that this definition, in contrast to what can sometimes be found in the literature, does
+    not require that $f(u, v) = -f(v, u)$. This extra requirement eliminates cycles between two
+    vertices and makes some theorem statements shorter. However, since larger cycles are still
+    permitted, this usually does not help in proofs. To the contrary, defining a flow becomes more
+    cumbersome, sometimes significantly so. -/
+@[ext]
+structure Network.Problem.Flow (Pr : N.Problem) where
+  /-- The assignment of flow units to each edge. -/
+  f : V → V → R
+  nonneg : ∀ u v, 0 ≤ f u v
+  conservation : ∀ v, v ≠ Pr.s ∧ v ≠ Pr.t → flowOut f v = flowIn f v
+  capacity : ∀ u v, f u v ≤ N.cap u v
+
+/-- The flow, which assigns zero to all edges. -/
+def Network.Problem.nullFlow (Pr : N.Problem) : Flow Pr where
+  f _ _ := 0
+  nonneg _ _ := le_rfl
+  conservation _ _ := rfl
+  capacity := N.nonneg
+
+instance {Pr : N.Problem} : Zero Pr.Flow where
+  zero := Pr.nullFlow
+
+end FlowDefinition
+
+namespace Network.Problem.Flow
+section FlowValue
+
+variable [OrderedAddCommGroup R] {N : Network V R} {Pr : N.Problem}
+
+/-- The value of a flow is the amount of flow units that leave the source vertex and arrive at the
+    sink vertex. In particular, this is equal to the outgoing flow of the source that does not
+    return to it again. -/
+def value (F : Flow Pr) := flowOut F.f Pr.s - flowIn F.f Pr.s
+
+@[simp]
+instance : Preorder (Flow Pr) := Preorder.lift Flow.value
+
+instance [IsTotal R LE.le] : IsTotalPreorder (Flow Pr) LE.le where
+  total F F' := by simp only [instPreorder, Preorder.lift, total_of]
+
+end FlowValue
+
+section Subflows
+
+variable [OrderedAddCommGroup R] {N : Network V R} {Pr : N.Problem}
+
+@[simp]
+instance : HasSubset (Flow Pr) where
+  Subset F₁ F₂ := F₁.f ≤ F₂.f
+
+instance : IsPartialOrder (Flow Pr) (· ⊆ ·) where
+  refl F := by simp only [instHasSubset, le_refl]
+  trans F₁ F₂ F₃ h₁₂ h₂₃ := by simp_all only [instHasSubset, le_trans h₁₂ h₂₃]
+  antisymm F₁ F₂ h₁₂ h₂₁ := by ext u v; exact le_antisymm (h₁₂ u v) (h₂₁ u v)
+
+@[simp]
+instance : HasSSubset (Flow Pr) where
+  SSubset F₁ F₂ := F₁.f < F₂.f
+
+instance : IsStrictOrder (Flow Pr) (· ⊂ ·) where
+  irrefl F := by simp only [instHasSSubset, lt_self_iff_false, not_false_eq_true, forall_const]
+  trans F₁ F₂ F₃ h₁₂ h₂₃ := by simp_all only [instHasSSubset, lt_trans h₁₂ h₂₃]
+
+instance : IsNonstrictStrictOrder (Flow Pr) (· ⊆ ·) (· ⊂ ·) where
+  right_iff_left_not_left F₁ F₂ := by
+    constructor <;> (intro h; simp_all only [instHasSSubset, instHasSubset]; exact h)
```

Comment:
Do you really really care about the subset notation? If not, I would suggest using `≤`/`<` instead as subset notation is a lot of overhead to maintain for little benefit (in fact, we are planning on moving away from it for `Set` and `Finset`).

Reply:
The existence of the subset relation itself is very useful, mostly for subtracting one flow from another. Since the resulting flow cannot have negative values for any edge, the definition in our seminar code currently is:
```lean
def Flow.sub {F₁ F₂ : Flow Pr} (h_le : F₁ ⊆ F₂) : Flow Pr where
  f := F₂.f - F₁.f
  -- ...
```

The second use case that occurred was for transferring "flow walks" between flows when decomposing a flow into paths: https://github.com/niklasmohrin/lean-seminar-2023/blob/5a53c8e067632017bded4fbf24b346b2ffddb84d/FlowEquivalentForest/Flow/Decomposition.lean#L484

---

The motivation for the subset notation was that if you imagine `F` as a collection of paths from s to t and cycles (or alternatively as a collection of edges), the subset relation would be exactly the multiset-subset relation among these collections. I suppose that in the context of mathlib however, this is exactly the opposite of how it works for `Multiset`.

I suppose an equally valid model would be to say that LE on flows is the same as LE on their underlying function `f` - this seems good to me.

The only benefit of the current system is that flows have order by their value, so we can denote a maximum flow with `IsTop F` or use `\top` to refer to a maximum flow. Is there any other existing definition for "maximal with respect to"?

Either way, I think you are right and only having LE, but defined like subset in the current version, seems to be the right move.

Reply:
Yes I think it's worth being consistent here and only use `≤`/`<` instead of `⊆`/`⊂`. Mind adapting the PR?